kvbc/src/categorization/versioned_kv_category.cpp

```diff
@@ -0,0 +1,347 @@
+                                         const VersionedDbValue &value,
+                                         storage::rocksdb::NativeWriteBatch &batch) {
+  batch.put(values_cf_, serialize(key), serialize(value));
+}
+
+void VersionedKeyValueCategory::addKeyToUpdateInfo(std::string &&key,
+                                                   bool deleted,
+                                                   bool stale_on_update,
+                                                   VersionedOutput &out) {
+  out.keys.emplace(std::move(key), VersionedKeyFlags{deleted, stale_on_update});
+}
+
+void VersionedKeyValueCategory::deleteGenesisBlock(BlockId block_id,
+                                                   const VersionedOutput &out,
+                                                   storage::rocksdb::NativeWriteBatch &batch) {
+  for (const auto &[key, flags] : out.keys) {
+    const auto latest_version = getLatestVersion(key);
+    ConcordAssert(latest_version.has_value());
+    if (flags.stale_on_update) {
+      // This key is marked stale-on-update and, therefore, we can remove its value at `block_id`.
+      batch.del(values_cf_, serialize(VersionedRawKey{key, block_id}));
+
+      // If there are no new versions of a key that was marked as stale-on-update at `block_id`, remove the versions
+      // index too.
+      if (*latest_version == block_id) {
+        batch.del(latest_ver_cf_, key);
+      }
+    } else if (*latest_version > block_id) {
+      // If this key is stale as of `block_id` (meaning it has a newer version), we can remove its value at `block_id`.
+      batch.del(values_cf_, serialize(VersionedRawKey{key, block_id}));
+    }
+  }
+}
+
+void VersionedKeyValueCategory::deleteLastReachableBlock(BlockId block_id,
+                                                         const VersionedOutput &out,
+                                                         storage::rocksdb::NativeWriteBatch &batch) {
+  for (const auto &[key, _] : out.keys) {
+    (void)_;
+    const auto versioned_key = serialize(VersionedRawKey{key, block_id});
+
+    // Find the previous version of the key and set it as a last version. Exploit the fact that CMF
+    // serializes strings prefixed by their length (in big-endian). Therefore, VersionedRawKeys will be ordered by key
+    // size, key and version. If not found, then this is the only version of the key and we can remove the latest
+    // version index too.
+    auto iter = db_->getIterator(values_cf_);
+    iter.seekAtMost(versioned_key);
+    ConcordAssert(iter);
+    iter.prev();
+    if (iter) {
+      auto prev_key = VersionedRawKey{};
+      deserialize(iter.keyView(), prev_key);
+      if (prev_key.value == key) {
+        updateLatestKeyVersion(key, prev_key.version, batch);
+      } else {
+        // This is the only version of the key - remove the latest version index too.
+        batch.del(latest_ver_cf_, key);
+      }
```

Comment:
Can it happen that you get an iterator that points to a different key (with the CMF exploit)? and if so don't we need a ```while``` to validate that no other keys might exist?

Reply:
I think you can get a different key when there is a single version of a key - you seek to it, go back and find some other key. The assumption is that equivalent keys are clustered together and ordered by version. Please let me know if you find a problem with that approach.